Rubberduck.Parsing/Binding/Bindings/IndexDefaultBinding.cs

```diff
@@ -123,178 +117,210 @@ private IBoundExpression CreateFailedExpression(IBoundExpression lExpression)
-                        }
-                        else if (defaultMember.DeclarationType == DeclarationType.Procedure)
-                        {
-                            classification = ExpressionClassification.Subroutine;
-                        }
-                        else
-                        {
-                            classification = ExpressionClassification.Function;
-                        }
-                        var defaultMemberAsLExpression = new SimpleNameExpression(defaultMember, classification, _expression);
-                        return Resolve(defaultMemberAsLExpression);
-                    }
+                /**
+                    This default member can accept no parameters. In this case, the static analysis restarts 
+                    recursively, as if this default member was specified instead for <l-expression> with the 
+                    same <argument-list>.
+                */
+                if (parameters.Count(parameter => !parameter.IsOptional) == 0
+                    && DEFAULT_MEMBER_RECURSION_LIMIT > defaultMemberResolutionRecursionDepth)
+                {
+                    return ResolveRecursiveDefaultMember(defaultMember, defaultMemberClassification, defaultMemberResolutionRecursionDepth);
                 }
             }
+
             return null;
         }
 
-        private IBoundExpression ResolveLExpressionDeclaredTypeIsArray(IBoundExpression lExpression, Declaration asTypeDeclaration)
+        private static bool ArgumentListIsCompatible(ICollection<ParameterDeclaration> parameters, ArgumentList argumentList)
         {
+            return (parameters.Count >= argumentList.Arguments.Count 
+                        || parameters.Any(parameter => parameter.IsParamArray))
+                    && parameters.Count(parameter => !parameter.IsOptional) <= argumentList.Arguments.Count;
+        }
+
+        private IBoundExpression ResolveRecursiveDefaultMember(Declaration defaultMember, ExpressionClassification defaultMemberClassification, int defaultMemberResolutionRecursionDepth)
+        {
+            var defaultMemberAsLExpression = new SimpleNameExpression(defaultMember, defaultMemberClassification, _expression);
+            return Resolve(defaultMemberAsLExpression, defaultMemberResolutionRecursionDepth + 1);
+        }
+
+        private static ExpressionClassification DefaultMemberExpressionClassification(Declaration defaultMember)
+        {
+            if (defaultMember.DeclarationType.HasFlag(DeclarationType.Property))
+            {
+                return ExpressionClassification.Property;
+            }
+
+            if (defaultMember.DeclarationType == DeclarationType.Procedure)
+            {
+                return ExpressionClassification.Subroutine;
+            }
+
+            return ExpressionClassification.Function;
+        }
+
+        private static bool IsPropertyGetLetFunctionProcedure(Declaration declaration)
```

Comment:
I note that this (and the following functions) seems to be also in the DictionaryAccessDefaultBinding. Can we DRY those helper functions?

Reply:
I would rather not like to DRY this. There is no good place to put these functions as they are of no use to anything else. Moreover, they do not contain any significant logic. The only reason they have been extracted is to make one conditional more readable.